refactor: add coinbase constraints to BlockAssembler::Options

Sjors opened the pull request:
When generating a block template through e.g. getblocktemplate RPC, we reserve 4000 weight units and 400 sigops. Pools use this space for their coinbase outputs.

At least one pool patched their Bitcoin Core node to adjust these hardcoded values. They eventually [produced an invalid block](https://bitcoin.stackexchange.com/questions/117837/how-many-sigops-are-in-the-invalid-block-783426) which exceeded the sigops limit.

The existince of such patches suggests it may be useful to make this value configurable. This PR would make such a change easier. However, the main motivation is that in the Stratum v2 spec requires the pool to communicate the maximum bytes they intend
to add to the coinbase outputs.

Specifically the `CoinbaseOutputDataSize` message which is part of the [Template Distribution Protocol](https://github.com/stratum-mining/sv2-spec/blob/main/07-Template-Distribution-Protocol.md#71-coinbaseoutputdatasize-client---server) has a field `coinbase_output_max_additional_size`.

A proposed change to the spec adds the max additional sigops as well: https://github.com/stratum-mining/sv2-spec/pull/86. Whether that change makes it into the spec is not important though, as adding both to `BlockAssembler::Options` makes sense.

The first commit is a test refactor followup for #30335, related to the code that's changed here, but not required.

The second commit introduces BlockCreateOptions, with just `use_mempool`.


The thirds commit adds `coinbase_max_additional_weight` and `coinbase_output_max_additional_sigops` to  `BlockCreateOptions`. They use the originally hardcoded values, and no existing caller overrides these defaults. This changes in #29432.

Sjors committed:
refactor: use CHECK_NONFATAL to avoid single-use symbol

tdb3 reviewed (commented):
Approach ACK
This adds flexibility for SV2, and having less magic numbers is better.  I support adding a configuration parameter for this (would prevent others from needing to fork).

Left a couple small nits for now, but plan to test/exercise in more detail.

TheCharlatan commented:
> I picked libbitcoin_util instead of libbitcoin_common to avoid a future dependency of the Stratum v2 Template Provider (https://github.com/bitcoin/bitcoin/pull/29432) on libbitcoin_common. The latter goal requires additional changes and is mostly just nice to have (the important bit is not depending on libbitcoin_node).

So I guess it should ideally only have the kernel as a dependency? It is not quite clear to me yet how this is going to be achieved. The current mining interface requires knowledge of a `CBlockTemplate`, which I don't think should be part of the kernel in the future.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -883,11 +883,24 @@ class MinerImpl : public Mining
         BlockAssembler::Options options;
         ApplyArgsManOptions(gArgs, options);
 
+        Assume(coinbase_output_max_additional_size <= DEFAULT_BLOCK_MAX_WEIGHT);
```
This probably needs to be `DEFAULT_BLOCK_MAX_WEIGHT / 4`, let me check the spec as well...

Sjors commented on src/interfaces/mining.h:
```diff
@@ -42,9 +43,13 @@ class Mining
      *
      * @param[in] script_pub_key the coinbase output
      * @param[in] use_mempool set false to omit mempool transactions
+     * @param[in] coinbase_output_max_additional_size maximum additional serialized bytes which the pool will add in coinbase transaction outputs
```
See also: https://github.com/stratum-mining/sv2-spec/blob/main/07-Template-Distribution-Protocol.md#71-coinbaseoutputdatasize-client---server

Sjors replied:
I ended up renaming `coinbase_output_max_additional_size` to `coinbase_max_additional_weight` to account for these issues, see below.

Sjors commented:
I renamed `coinbase_output_max_additional_size` to `coinbase_max_additional_weight`. With the current [Stratum v2 spec](https://github.com/stratum-mining/sv2-spec/blob/main/07-Template-Distribution-Protocol.md#71-coinbaseoutputdatasize-client---server) the latter is calculated as follows:

`coinbase_max_additional_weight = (coinbase_output_max_additional_size + 100 + 0 + 2) * 4`

- `coinbase_output_max_additional_size * 4`: Coinbase outputs are not part of the witness so their weight is simply the number of bytes times 4.

- `100 * 4`: the spec also requires taking into account a "maximally-sized (100 byte) coinbase field", which refers to the scriptSig of the coinbase. It currently doesn't allow for setting a custom coinbase witness.

- `0 * 4`: a [CompactSize](https://bitcoin.stackexchange.com/a/114585/4948) encodes the size of the above "coinbase field" (the spec is ambiguous whether this is in included). Given the maximum of 100 its always encoded with 1 byte, so no _additional_ weight units are needed.

- `2 * 4`: "Further, the Template Provider MUST consider the maximum additional bytes required in the output count variable-length integer in the coinbase transaction when complying with the size limits." - this is tricky because we don't know how many outputs the pool intends to add, and we might want to add an arbitrary number of outputs ourselves. Since there can't be more than 1 million outputs, 3 bytes is always enough to encode the number of outputs. So that's 2 _additional_ bytes in the worst case.

The spec should probably be clarified a bit, cc @TheBlueMatt, @Fi3.

In any case, tracking coinbase reserved space in term of weight units here should be enough to support Stratum v2 later, and even allow for using the coinbase witness.

Sjors commented:
Rebased after #30324.

TheCharlatan commented:
Is there a reason why
```
BlockAssembler::Options options;
ApplyArgsManOptions(gArgs, options);
```
has to be inside the block assembler class? Why not construct them externally and pass a reference of the options to the `BlockAssembler`? This seems a bit more efficient, since the options don't have to be constructed and the arguments not read every time a template is retrieved.

Sjors commented:
cc @glozow who added this as part of #26695.

glozow commented:
> Is there a reason why
> 
> ```
> BlockAssembler::Options options;
> ApplyArgsManOptions(gArgs, options);
> ```
> 
> has to be inside the block assembler class?

It doesn't need to be inside `BlockAssembler`. Prior to #26695, the ctor was using the global `gArgs` to decide on parameters (yuck!). We didn't have the node/*args kernel/*opts conventions at the time and I was pretty happy adding an `ApplyArgsMan` that didn't do that.

> Why not construct them externally and pass a reference of the options to the `BlockAssembler`?

That sounds fine.

luke-jr commented:
Weak approach NACK. This is just limiting the max block weight, which is already a configurable parameter. Having two params for the same thing seems like a poor way to do it. I'm not sure it avoids the risk of bugs producing invalid blocks, either - it just moves the bug risk from one place to another equally probable.

Instead, maybe just perform the block weight cap on initialization rather than block assembly?

Sjors commented:
> This is just limiting the max block weight

This had me confused initially, and I tried to combine them, but there's really two different things:

1. The max block weight demanded by the user via `-blockmaxweight`
2. Weight units reserved for the coinbase

With stratum v1 the distinction isn't very important. We just subtract (2) from (1). We could indeed do that at initialization.

But with Stratum v2 (2) is a per-client setting communicated to us (the Template Provider) via the [CoinbaseOutputDataSize](https://github.com/stratum-mining/sv2-spec/blob/main/07-Template-Distribution-Protocol.md) message. We can set it at connection time, but not during node initialization.

>> Why not construct them externally and pass a reference of the options to the BlockAssembler?

> That sounds fine.

Happy to take patches.

itornaza commented on src/interfaces/mining.h:
```diff
@@ -42,9 +43,13 @@ class Mining
      *
      * @param[in] script_pub_key the coinbase output
      * @param[in] use_mempool set false to omit mempool transactions
+     * @param[in] coinbase_max_addition_weight maximum additional weight which the pool will add to the coinbase transaction
```
I take it you wanted to rename it to `coinbase_max_additional_weight` instead

Sjors replied:
Thanks.

itornaza commented on src/interfaces/mining.h:
```diff
@@ -42,9 +43,13 @@ class Mining
      */
-    virtual std::unique_ptr<node::CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool = true) = 0;
+    virtual std::unique_ptr<node::CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool = true,
+                                                                 size_t coinbase_max_addition_weight = DEFAULT_COINBASE_MAX_ADDITIONAL_WEIGHT,
```
Same here, did you meant to name it `coinbase_max_additional_weight`?

itornaza commented on src/node/interfaces.cpp:
```diff
@@ -883,11 +883,24 @@ class MinerImpl : public Mining
 
-    std::unique_ptr<CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool) override
+    std::unique_ptr<CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool,
+                                                   size_t coinbase_max_addition_weight,
```
Again, should it be `coinbase_max_additional_weight?`?

itornaza commented on src/node/interfaces.cpp:
```diff
@@ -883,11 +883,24 @@ class MinerImpl : public Mining
         ApplyArgsManOptions(gArgs, options);
 
+        Assume(coinbase_max_addition_weight <= DEFAULT_BLOCK_MAX_WEIGHT);
+        options.coinbase_max_additional_weight = coinbase_max_addition_weight;
```
`coinbase_max_additional_weight`?

itornaza reviewed (approved):
Concept ACK d89637eceab0145a64c2d1af1139ace204a3ab3c

Giving the option to assign the values of `coinbase_max_additional_weight` and `coinbase_output_max_additional_sigops` needed by the Stratum V2 protocol is essential for its integration. I understand the risks involved by having two parameters for the same thing:

>     1. The max block weight demanded by the user via -blockmaxweight
>     2. Weight units reserved for the coinbase

but it seems unavoidable at a first glance due to their different invocation domains. During the node initialisation for the former and connection time for the later.

Left some small nit that only applies if the intended name for the variable was `coinbase_max_additional_weight` instead of `coinbase_max_addition_weight`

ryanofsky commented:
Code review ACK e67e4669a9e263d90f0e276c8e65e93c39170375, but I think the current implementation is too complicated and has too many layers of indirection.

Would suggest applying the following patch and squashing the PR down to a single commit which would be just +41/-16 with the suggested changes:

<details><summary>diff</summary>
<p>

```diff
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -313,7 +313,6 @@ BITCOIN_CORE_H = \
   util/hasher.h \
   util/insert.h \
   util/macros.h \
-  util/mining.h \
   util/moneystr.h \
   util/overflow.h \
   util/overloaded.h \
--- a/src/interfaces/mining.h
+++ b/src/interfaces/mining.h
@@ -5,10 +5,11 @@
 #ifndef BITCOIN_INTERFACES_MINING_H
 #define BITCOIN_INTERFACES_MINING_H
 
+#include <node/types.h>
+#include <uint256.h>
+
 #include <memory>
 #include <optional>
-#include <uint256.h>
-#include <util/mining.h>
 
 namespace node {
 struct CBlockTemplate;
@@ -42,14 +43,10 @@ public:
      * Construct a new block template
      *
      * @param[in] script_pub_key the coinbase output
-     * @param[in] use_mempool set false to omit mempool transactions
-     * @param[in] coinbase_max_additional_weight maximum additional weight which the pool will add to the coinbase transaction
-     * @param[in] coinbase_output_max_additional_sigops maximum additional sigops which the pool will add in coinbase transaction outputs
+     * @param[in] options options for creating the block
      * @returns a block template
      */
-    virtual std::unique_ptr<node::CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool = true,
-                                                                 size_t coinbase_max_additional_weight = DEFAULT_COINBASE_MAX_ADDITIONAL_WEIGHT,
-                                                                 size_t coinbase_output_max_additional_sigops = DEFAULT_COINBASE_OUTPUT_MAX_ADDITIONAL_SIGOPS) = 0;
+    virtual std::unique_ptr<node::CBlockTemplate> createNewBlock(const CScript& script_pub_key, const node::BlockCreateOptions& options={}) = 0;
 
     /**
      * Processes new block. A valid new block is automatically relayed to peers.
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -883,25 +883,11 @@ public:
         return TestBlockValidity(state, chainman().GetParams(), chainman().ActiveChainstate(), block, tip, /*fCheckPOW=*/false, check_merkle_root);
     }
 
-    std::unique_ptr<CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool,
-                                                   size_t coinbase_max_additional_weight,
-                                                   size_t coinbase_output_max_additional_sigops) override
+    std::unique_ptr<CBlockTemplate> createNewBlock(const CScript& script_pub_key, const BlockCreateOptions& options) override
     {
-        BlockAssembler::Options options;
-        ApplyArgsManOptions(gArgs, options);
-
-        Assume(coinbase_max_additional_weight <= DEFAULT_BLOCK_MAX_WEIGHT);
-        options.coinbase_max_additional_weight = coinbase_max_additional_weight;
-
-        Assume(options.coinbase_output_max_additional_sigops <= MAX_BLOCK_SIGOPS_COST);
-        options.coinbase_output_max_additional_sigops = coinbase_output_max_additional_sigops;
-
-        // The BlockAssembler constructor calls ClampOptions which clamps
-        // nBlockMaxWeight between coinbase_output_max_additional_size and
-        // DEFAULT_BLOCK_MAX_WEIGHT.
-        // In other words, coinbase (reserved) outputs can safely exceed
-        // -blockmaxweight, but the rest of the block template will be empty.
-        return BlockAssembler{chainman().ActiveChainstate(), use_mempool ? context()->mempool.get() : nullptr, options}.CreateNewBlock(script_pub_key);
+        BlockAssembler::Options assemble_options{options};
+        ApplyArgsManOptions(*Assert(m_node.args), assemble_options);
+        return BlockAssembler{chainman().ActiveChainstate(), context()->mempool.get(), assemble_options}.CreateNewBlock(script_pub_key);
     }
 
     NodeContext* context() override { return &m_node; }
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -59,14 +59,17 @@ void RegenerateCommitments(CBlock& block, ChainstateManager& chainman)
 
 static BlockAssembler::Options ClampOptions(BlockAssembler::Options options)
 {
+    Assume(options.coinbase_max_additional_weight <= DEFAULT_BLOCK_MAX_WEIGHT);
+    Assume(options.coinbase_output_max_additional_sigops <= MAX_BLOCK_SIGOPS_COST);
     // Limit weight to between coinbase_max_additional_weight and DEFAULT_BLOCK_MAX_WEIGHT for sanity:
+    // Coinbase (reserved) outputs can safely exceed -blockmaxweight, but the rest of the block template will be empty.
     options.nBlockMaxWeight = std::clamp<size_t>(options.nBlockMaxWeight, options.coinbase_max_additional_weight, DEFAULT_BLOCK_MAX_WEIGHT);
     return options;
 }
 
 BlockAssembler::BlockAssembler(Chainstate& chainstate, const CTxMemPool* mempool, const Options& options)
     : chainparams{chainstate.m_chainman.GetParams()},
-      m_mempool{mempool},
+      m_mempool{options.use_mempool ? mempool : nullptr},
       m_chainstate{chainstate},
       m_options{ClampOptions(options)}
 {
--- a/src/node/miner.h
+++ b/src/node/miner.h
@@ -6,6 +6,7 @@
 #ifndef BITCOIN_NODE_MINER_H
 #define BITCOIN_NODE_MINER_H
 
+#include <node/types.h>
 #include <policy/policy.h>
 #include <primitives/block.h>
 #include <txmempool.h>
@@ -20,8 +21,6 @@
 #include <boost/multi_index/tag.hpp>
 #include <boost/multi_index_container.hpp>
 
-#include <util/mining.h>
-
 class ArgsManager;
 class CBlockIndex;
 class CChainParams;
@@ -155,24 +154,13 @@ private:
     Chainstate& m_chainstate;
 
 public:
-    struct Options {
+    struct Options : BlockCreateOptions {
         // Configuration parameters for the block size
         size_t nBlockMaxWeight{DEFAULT_BLOCK_MAX_WEIGHT};
         CFeeRate blockMinFeeRate{DEFAULT_BLOCK_MIN_TX_FEE};
         // Whether to call TestBlockValidity() at the end of CreateNewBlock().
         bool test_block_validity{true};
         bool print_modified_fee{DEFAULT_PRINT_MODIFIED_FEE};
-        /**
-         * The maximum additional weight which the pool will add to the coinbase
-         * scriptSig, witness and outputs. This must include any additional
-         * weight needed for larger CompactSize encoded lengths.
-         */
-        size_t coinbase_max_additional_weight{DEFAULT_COINBASE_MAX_ADDITIONAL_WEIGHT};
-        /**
-         * The maximum additional sigops which the pool will add in coinbase
-         * transaction outputs.
-         */
-        size_t coinbase_output_max_additional_sigops{DEFAULT_COINBASE_OUTPUT_MAX_ADDITIONAL_SIGOPS};
     };
 
     explicit BlockAssembler(Chainstate& chainstate, const CTxMemPool* mempool, const Options& options);
--- a/src/node/types.h
+++ b/src/node/types.h
@@ -13,6 +13,8 @@
 #ifndef BITCOIN_NODE_TYPES_H
 #define BITCOIN_NODE_TYPES_H
 
+#include <cstddef>
+
 namespace node {
 enum class TransactionError {
     OK, //!< No error
@@ -24,6 +26,24 @@ enum class TransactionError {
     MAX_BURN_EXCEEDED,
     INVALID_PACKAGE,
 };
+
+struct BlockCreateOptions {
+    /**
+     * Set false to omit mempool transactions
+     */
+    bool use_mempool{true};
+    /**
+     * The maximum additional weight which the pool will add to the coinbase
+     * scriptSig, witness and outputs. This must include any additional
+     * weight needed for larger CompactSize encoded lengths.
+     */
+    size_t coinbase_max_additional_weight{4000};
+    /**
+     * The maximum additional sigops which the pool will add in coinbase
+     * transaction outputs.
+     */
+    size_t coinbase_output_max_additional_sigops{400};
+};
 } // namespace node
 
 #endif // BITCOIN_NODE_TYPES_H
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -371,7 +371,7 @@ static RPCHelpMan generateblock()
 
     ChainstateManager& chainman = EnsureChainman(node);
     {
-        std::unique_ptr<CBlockTemplate> blocktemplate{miner.createNewBlock(coinbase_script, /*use_mempool=*/false)};
+        std::unique_ptr<CBlockTemplate> blocktemplate{miner.createNewBlock(coinbase_script, {.use_mempool = false})};
         if (!blocktemplate) {
             throw JSONRPCError(RPC_INTERNAL_ERROR, "Couldn't create new block");
         }
deleted file mode 100644
--- a/src/util/mining.h
+++ /dev/null
@@ -1,13 +0,0 @@
-// Copyright (c) 2024 The Bitcoin Core developers
-// Distributed under the MIT software license, see the accompanying
-// file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
-#ifndef BITCOIN_UTIL_MINING_H
-#define BITCOIN_UTIL_MINING_H
-
-/** Default for coinbase_max_additional_weight */
-static constexpr unsigned int DEFAULT_COINBASE_MAX_ADDITIONAL_WEIGHT{4000};
-/** Default for coinbase_output_max_additional_sigops */
-static constexpr unsigned int DEFAULT_COINBASE_OUTPUT_MAX_ADDITIONAL_SIGOPS{400};
-
-#endif // BITCOIN_UTIL_MINING_H
```

</p>
</details>

Suggested changes:

- Make `createNewBlock` use an options struct so it easier and safer to use (less likely to have bugs from mixing up the order of integer parameters).
- Consolidates documentation for the options in one header instead of three headers.
- Avoids adding `util/mining.h` header since the options more logically belong to the node library than the util library.
- Moves implementation of mining options from `node/interfaces.cpp` to `node/miner.cpp`. The `interfaces.cpp` file is just meant to contain glue code exposing private implementations as public interfaces. Ideally, the interface methods should be 1-3 lines long and not implement real functionality, which belongs in other places like `validation.cpp` and `miner.cpp` so it can be found and reused more easily and better tested.

ismaelsadeeq commented on src/node/miner.h:
```diff
@@ -160,6 +160,17 @@ class BlockAssembler
+         * The maximum additional sigops which the pool will add in coinbase
+         * transaction outputs.
+         */
+        size_t coinbase_output_max_additional_sigops{400};
```
I think it should not be necessary to reference pool in here

```suggestion
        /**
         * The maximum additional weight that can be added  to the coinbase
         * scriptSig, witness and outputs. This must include any additional
         * weight needed for larger CompactSize encoded lengths.
         */
        size_t coinbase_max_additional_weight{4000};
        /**
         * The maximum additional sigops that can be added in coinbase
         * transaction outputs.
         */
        size_t coinbase_output_max_additional_sigops{400};
```

Sjors replied:
See above https://github.com/bitcoin/bitcoin/pull/30356#discussion_r1660679371, just omitting the word "pool" because in theory there could be something _other_ than a pool that wants to add data to our coinbase, just seems confusing to me.

ismaelsadeeq commented on src/node/miner.h:
```diff
@@ -160,6 +160,17 @@ class BlockAssembler
+         * The maximum additional weight which the pool will add to the coinbase
+         * scriptSig, witness and outputs. This must include any additional
+         * weight needed for larger CompactSize encoded lengths.
+         */
```
Indicate the assumed upper limits of this options?

Sjors replied:
#29432 implements this in more detail. I'm not even 100% sure if it's correct there, so I'd rather wait with documentation the exact numbers - that might just encourage someone to recompile and burn themselves.

ismaelsadeeq commented on src/node/interfaces.cpp:
```diff
@@ -883,11 +883,24 @@ class MinerImpl : public Mining
         BlockAssembler::Options options;
         ApplyArgsManOptions(gArgs, options);
 
+        Assume(coinbase_max_additional_weight <= DEFAULT_BLOCK_MAX_WEIGHT);
+        options.coinbase_max_additional_weight = coinbase_max_additional_weight;
+
+        Assume(options.coinbase_output_max_additional_sigops <= MAX_BLOCK_SIGOPS_COST);
```
Why are we allowing for this large values?
shouldnt this be 
```suggestion
         Assume(coinbase_max_additional_weight <= MAX_STANDARD_TX_WEIGHT);
    options.coinbase_max_additional_weight = coinbase_max_additional_weight;
     Assume(options.coinbase_output_max_additional_sigops <= MAX_STANDARD_TX_SIGOPS_COST);
```

Sjors replied:
`MAX_STANDARD_TX_WEIGHT` is a standardness rule. If we wanted to enforce it in the mining code, which would be unrelated to this PR, it shouldn't be through an assert / assume.

Sjors commented:
@ryanofsky thanks! I'll look into using your patch. After that I'll check where @ismaelsadeeq's feedback still applies.

Sjors committed:
refactor: pass BlockCreateOptions to createNewBlock

Rather than pass options individually to createNewBlock and then
combining them into BlockAssembler::Options, this commit introduces
BlockCreateOptions and passes that instead.

Currently there's only one option (use_mempool) but the next
commit adds more.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

Sjors commented:
I took @ryanofsky's patch and split it into two commits.

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -59,8 +59,11 @@ void RegenerateCommitments(CBlock& block, ChainstateManager& chainman)
 {
-    // Limit weight to between 4K and DEFAULT_BLOCK_MAX_WEIGHT for sanity:
-    options.nBlockMaxWeight = std::clamp<size_t>(options.nBlockMaxWeight, 4000, DEFAULT_BLOCK_MAX_WEIGHT);
+    Assume(options.coinbase_max_additional_weight <= DEFAULT_BLOCK_MAX_WEIGHT);
+    Assume(options.coinbase_output_max_additional_sigops <= MAX_BLOCK_SIGOPS_COST);
```
In commit "Pass coinbase constraints to createNewBlock" (e67e4669a9e263d90f0e276c8e65e93c39170375)

Current code seems ok, but depending on how important it is for this not to generate an invalid block, it may make sense to use `assert` instead of `Assume`, since Assume does not do anything in release builds. Or this could also abort in debug builds but cap the values in release builds:

```c++
if (!Assume(options.coinbase_output_max_additional_sigops <= MAX_BLOCK_SIGOPS_COST)) {
    options.coinbase_output_max_additional_sigops = MAX_BLOCK_SIGOPS_COST
}
```

Sjors replied:
Currently neither of these values can be configured. With Stratum v2 I intend to add runtime checks. If those checks are wrong then I assume the fuzzer would eventually find this `Assume`?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30356#discussion_r1679851292

So you are saying if the values are too big at this point, it means there is a bug elsewhere in the code. If that is the case, I think it would be more appropriate to use `assert` or `Assert` rather than `Assume` so the bug does not go undetected until it is too late and eventually results in an invalid block being produced. I could be wrong, but I don't see a rationale for preferring assume over assert here.

Sjors replied:
> So you are saying if the values are too big at this point, it means there is a bug elsewhere in the code

Yes

Changed to Assert.

itornaza replied:
This is great! Despite that these values are hard coded, I also think it is safer to prefer an assert over Assume in this case. In that way we still have a safeguard in place during the whole integration process of Stratum V2 and even if the future run time checks somehow fail to deliver their intended purpose at least we have some feedback in debug mode. On the other hand the assert wont hurt anyone exactly because the values are hard coded.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30356#discussion_r1681361423

Just to clarify for other reviewers since the names of these macros are confusing: In bitcoin core, `assert` and `Assert` macros behave very differently from the `assert` macro in standard c++. In standard c++ `assert` statements are not executed at all in release builds, while in bitcoin core they are always executed and always cause the execution to abort if they fail. Other c++ libraries that don't to be confusing use the term [check](https://github.com/bitcoin/bitcoin/pull/16136#issuecomment-501779934) instead of assert for this behavior.

The `Assume` macro is more unusual, because it doesn't abort execution in release builds, but does also doesn't get skipped in release builds, so will always evaluate arguments passed to it. It can make sense to use Assume instead of assert when an unexpected condition indicates some kind of bug, but the bug is easy to recover from, so it is worth crashing debug builds but not release builds. Normally it make sense to check the `Assume` return value and do something like log a warning or fix invalid values, not just ignore the unexpected condition completely.

ryanofsky reviewed (approved):
Code review ACK 95c2edefe383c2b8a1735c26d8442fd36fa5e339

tdb3 reviewed (approved):
Code Review ACK 95c2edefe383c2b8a1735c26d8442fd36fa5e339
Good addition and the updates tidy things up nicely.

I would also support a bitcoind config parameter or possibly an RPC for this.

ryanofsky reviewed (approved):
Code review ACK 95c2edefe383c2b8a1735c26d8442fd36fa5e339

Sjors committed:
refactor: add coinbase constraints to BlockCreateOptions

When generating a block template through e.g. getblocktemplate RPC,
we reserve 4000 weight units and 400 sigops. Pools use this space
for their coinbase outputs.

At least one pool patched their Bitcoin Core node to adjust
these hardcoded values. They eventually produced an invalid
block which exceeded the sigops limit.
https://bitcoin.stackexchange.com/questions/117837/how-many-sigops-are-in-the-invalid-block-783426

The existince of such patches suggests it may be useful to
make this value configurable. This commit would make such a
change easier.

The main motivation however is that the Stratum v2 spec
requires the pool to communicate the maximum bytes they intend
to add to the coinbase outputs. A proposed change to the spec
would also require them to communicate the maximum number of sigops.

This commit also documents what happens when
-blockmaxweight is lower than the coinbase
reserved value.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

itornaza reviewed (approved):
tested ACK c504b6997b1acc9771ad1f52efaa4be2b4966c6c

Reviewed all changes since my last https://github.com/bitcoin/bitcoin/pull/30356#pullrequestreview-2167070268. Built the commit and run all unit, functional and extended tests and all of them pass.

ryanofsky reviewed (approved):
Code review ACK c504b6997b1acc9771ad1f52efaa4be2b4966c6c

ismaelsadeeq reviewed (approved):
Code review ACK c504b6997b1acc9771ad1f52efaa4be2b4966c6c

The PR description needs to be updated to match the PR.

Left a single question

Sjors commented:
@ismaelsadeeq thanks, updated the description.

Sjors commented on src/node/types.h:
```diff
@@ -24,6 +26,13 @@ enum class TransactionError {
+     * Set false to omit mempool transactions
+     */
+    bool use_mempool{true};
+};
```
No strong reason, but the coinbase script is different for every block, unlike the other stuff we pass to `BlockCreateOptions`. It seems worthy of a positional argument.

glozow commented:
post merge ACK
